src/convopt.hpp

```diff
@@ -10,25 +10,26 @@
 
 namespace opt {
 
-// TODO(enrlov): F must be invocable with T
-template <std::size_t N, class T, class F>
-constexpr auto gradient(point<T, N> p, F cost) -> distance_t<point<T, N>>
-{
-    // TODO(olee): Should point also be a range?
-    // TODO(olee): Should point also be sized?
-    // TODO(olee): Should point be rebindable?
+// TODO(enrlov): F must be invocable with P and opt::point<dual<scalar_t<P>>>
+// and return scalar_t<...> of the input
 
-    // using T = scalar_t<distance_t<decltype(p)>>;
+template <Point P, class F>
+constexpr auto gradient(const P& p, F cost) -> distance_t<P>
```

Comment:
The most wonderful return here would tangent_vector (and all the concepts and the algebra behind manifolds). But that would add an enormous layer of complexity.

Reply:
Would `tangent_vector` differ from the vector type returned by `distance_t<P>`?

Reply:
No, it would just express the same concept in a more generic (the more generic? I don't know enough to say that) way, the tangent vector is a vector in an Euclidean flat space